examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BdxOtaSender.h"
+
+#include <core/CHIPError.h>
+#include <messaging/ExchangeContext.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <support/BitFlags.h>
+
+#include <fstream>
+
+#include <string.h>
+
+using namespace chip::bdx;
+
+void BdxOtaSender::SetFilepath(const char * path, size_t length)
+{
+    if (path != nullptr)
+    {
+        strcpy(mFilepath, path);
+    }
+    else
+    {
+        memset(mFilepath, 0, kFilepathMaxLength);
+    }
+}
+
+void BdxOtaSender::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (event.EventType != TransferSession::OutputEventType::kNone)
+    {
+        ChipLogDetail(BDX, "OutputEvent type: %d", static_cast<uint16_t>(event.EventType));
+    }
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kNone:
+        break;
+    case TransferSession::OutputEventType::kMsgToSend:
+        VerifyOrReturn(mExchangeCtx != nullptr, ChipLogError(BDX, "%s: mExchangeCtx is null", __FUNCTION__));
+        err = mExchangeCtx->SendMessage(event.msgTypeData.ProtocolId, event.msgTypeData.MessageType, std::move(event.MsgData));
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(BDX, "SendMessage failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    case TransferSession::OutputEventType::kInitReceived: {
+        // TransferSession will automatically reject a transfer if there are no
+        // common supported control modes. It will also default to the smaller
+        // block size.
+        TransferSession::TransferAcceptData acceptData;
+        acceptData.ControlMode  = TransferControlFlags::kReceiverDrive; // OTA must use receiver drive
+        acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+        acceptData.StartOffset  = mTransfer.GetStartOffset();
+        acceptData.Length       = mTransfer.GetTransferLength();
+        VerifyOrReturn(mTransfer.AcceptTransfer(acceptData) == CHIP_NO_ERROR,
+                       ChipLogError(BDX, "%s: %s", __FUNCTION__, chip::ErrorStr(err)));
+        break;
+    }
+    break;
```

Comment:
These breaks seem off - probably should be deleted.

Reply:
Good catch, not sure where those came from.